source/numerics/matrix_tools.cc

```diff
@@ -2559,12 +2557,6 @@ namespace MatrixTools
             Assert (local_range == solution.local_range(),
                     ExcInternalError());
 
-            // we have to read and write from this
-            // matrix (in this order). this will only
-            // work if we compress the matrix first,
-            // done here
-            matrix.compress ();
-
```

Comment:
Same here. I think this was particularly insidious: it happens in a block guarded by `if (boundary_values.size()>0)` which means that interior processors did not call the function and it would probably have resulted in a deadlock in that case.